### ISSUE TITLE
Segfault in InitRHI when creating textures on Linux and macOS

### ISSUE THREAD

**Issue by maumock** [CesiumGS/cesium-unreal#1007]
Small issues with tag v1.20.0-ue5.
1. generated lib names have "d.a" instead of ".a" in their name
2. crash when adding cesium terrain

Happy to test any hot fixes or patches if needed.

TLDR

**Machine**

- unreal engine code base: 5.0.3-release
- Ubuntu 22.04 (clean install) also tried latest pop os
- Wayland
- RTX 3060 12gb 
- AMD 12 core
- 32gb ram

**Lib Issue**
On compile all the libraries in cesium-unreal (Source/ThirdParty/lib/Linux-x64) have "d.a" in the name (i.e. libCesiumGeometryd.a) instead of ".a" (i.e. libCesiumGeometry.a). Obviously this creates problems when linking. After renaming the lib files, compilation will work. 

**Crash**
When clicking on "Cesium World Terrain + Bing Maps World imagery", a crash occurs immediately. with the following error: 
```
Caught signal 11 Segmentation fault

libUnrealEditor-CesiumRuntime.so!(anonymous namespace)::FCesiumTextureResource::InitRHI() [/home/rho/dev/packages/CesiumForUnreal/HostProject/Plugins/CesiumForUnreal/Source/CesiumRuntime/Private/CesiumTextureUtility.cpp:299]
libUnrealEditor-RenderCore.so!FRenderResource::InitResource() [/home/rho/dev/UnrealEngine/Engine/Source/./Runtime/RenderCore/Private/RenderResource.cpp:120]
libUnrealEditor-CesiumRuntime.so!TEnqueueUniqueRenderCommandType<CesiumTextureUtility::loadTextureGameThreadPart(CesiumTextureUtility::LoadedTextureResult*)::Cesium_InitResourceName, CesiumTextureUtility::loadTextureGameThreadPart(CesiumTextureUtility::LoadedTextureResult*)::$_125>::DoTask(ENamedThreads::Type, TRefCountPtr<FGraphEvent> const&) [/home/rho/dev/UnrealEngine/Engine/Source/Runtime/RenderCore/Public/RenderingThread.h:193]
libUnrealEditor-**CesiumRuntime.so!TGraphTask<TEnqueueUniqueRenderCommandType<CesiumTextureUtility::loadTextureGameThreadPart(CesiumTextureUtility::LoadedTextureResult*)::Cesium_InitResourceName, CesiumTextureUtility::loadTextureGameThreadPart(CesiumTextureUtility::LoadedTextureResult*)::$_125>** >::ExecuteTask(TArray<FBaseGraphTask*, TSizedDefaultAllocator<32> >&, ENamedThreads::Type, bool) [/home/rho/dev/UnrealEngine/Engine/Source/Runtime/Core/Public/Async/TaskGraphInterfaces.h:975]
libUnrealEditor-Core.so!FNamedTaskThread::ProcessTasksNamedThread(int, bool) [/home/rho/dev/UnrealEngine/Engine/Source/./Runtime/Core/Private/Async/TaskGraph.cpp:753]
libUnrealEditor-Core.so!FNamedTaskThread::ProcessTasksUntilQuit(int) [/home/rho/dev/UnrealEngine/Engine/Source/./Runtime/Core/Private/Async/TaskGraph.cpp:641]
libUnrealEditor-Core.so!FTaskGraphCompatibilityImplementation::ProcessThreadUntilRequestReturn(ENamedThreads::Type) [/home/rho/dev/UnrealEngine/Engine/Source/./Runtime/Core/Private/Async/TaskGraph.cpp:2114]
libUnrealEditor-RenderCore.so!RenderingThreadMain(FEvent*) [/home/rho/dev/UnrealEngine/Engine/Source/./Runtime/RenderCore/Private/RenderingThread.cpp:380]
libUnrealEditor-RenderCore.so!FRenderingThread::Run() [/home/rho/dev/UnrealEngine/Engine/Source/./Runtime/RenderCore/Private/RenderingThread.cpp:544]
libUnrealEditor-Core.so!FRunnableThreadPThread::Run() [/home/rho/dev/UnrealEngine/Engine/Source/./Runtime/Core/Private/HAL/PThreadRunnableThread.cpp:25]
libUnrealEditor-Core.so!FRunnableThreadPThread::_ThreadProc(void*) [/home/rho/dev/UnrealEngine/Engine/Source/Runtime/Core/Private/HAL/PThreadRunnableThread.h:185]
libc.so.6!UnknownFunction(0x94b42)
libc.so.6!UnknownFunction(0x1269ff)
```

**Comment by maumock** [comment]
**Lib Issue**
The extern/CMakeLists.txt has the following settings for debug builds:
`set(CESIUM_DEBUG_POSTFIX "d")`

The setting appends a "d" to the file name. This causes issues if building due to lib deps having the incorrect name for linking. I assume docs could be adjusted to warn the user or plugin files could be updated if using a debug build.

**Comment by maumock** [comment]
Kudos for getting the v1.20-ue5 branch to build against 5.1. I really appreciate it!

**Crash**
Recompiled against a clean build of ue 5.1.0-release and had the same seg fault in cesium texture util:

`libUnrealEditor-CesiumRuntime.so!(anonymous namespace)::FCesiumTextureResource::InitRHI() [/home/rho/dev/packages/CesiumForUnreal/HostProject/Plugins/CesiumForUnreal/Source/CesiumRuntime/Private/CesiumTextureUtility.cpp:299]
libUnrealEditor-RenderCore.so!FRenderResource::InitResource() [/home/rho/dev/UnrealEngine/Engine/Source/./Runtime/RenderCore/Private/RenderResource.cpp:277]
libUnrealEditor-CesiumRuntime.so!TEnqueueUniqueRenderCommandType<CesiumTextureUtility::loadTextureGameThreadPart(CesiumTextureUtility::LoadedTextureResult*)::Cesium_InitResourceName, CesiumTextureUtility::loadTextureGameThreadPart(CesiumTextureUtility::LoadedTextureResult*)::$_108>::DoTask(ENamedThreads::Type, TRefCountPtr<FGraphEvent> const&) [/home/rho/dev/UnrealEngine/Engine/Source/Runtime/RenderCore/Public/RenderingThread.h:206]
libUnrealEditor-CesiumRuntime.so!TGraphTask<TEnqueueUniqueRenderCommandType<CesiumTextureUtility::loadTextureGameThreadPart(CesiumTextureUtility::LoadedTextureResult*)::Cesium_InitResourceName, CesiumTextureUtility::loadTextureGameThreadPart(CesiumTextureUtility::LoadedTextureResult*)::$_108> >::ExecuteTask(TArray<FBaseGraphTask*, TSizedDefaultAllocator<32> >&, ENamedThreads::Type, bool) [/home/rho/dev/UnrealEngine/Engine/Source/Runtime/Core/Public/Async/TaskGraphInterfaces.h:1348]
libUnrealEditor-Core.so!FNamedTaskThread::ProcessTasksNamedThread(int, bool) [/home/rho/dev/UnrealEngine/Engine/Source/./Runtime/Core/Private/Async/TaskGraph.cpp:760]
libUnrealEditor-Core.so!FNamedTaskThread::ProcessTasksUntilQuit(int) [/home/rho/dev/UnrealEngine/Engine/Source/./Runtime/Core/Private/Async/TaskGraph.cpp:648]
libUnrealEditor-Core.so!FTaskGraphCompatibilityImplementation::ProcessThreadUntilRequestReturn(ENamedThreads::Type) [/home/rho/dev/UnrealEngine/Engine/Source/./Runtime/Core/Private/Async/TaskGraph.cpp:2149]
libUnrealEditor-RenderCore.so!RenderingThreadMain(FEvent*) [/home/rho/dev/UnrealEngine/Engine/Source/./Runtime/RenderCore/Private/RenderingThread.cpp:415]
libUnrealEditor-RenderCore.so!FRenderingThread::Run() [/home/rho/dev/UnrealEngine/Engine/Source/./Runtime/RenderCore/Private/RenderingThread.cpp:566]
libUnrealEditor-Core.so!FRunnableThreadPThread::Run() [/home/rho/dev/UnrealEngine/Engine/Source/./Runtime/Core/Private/HAL/PThreadRunnableThread.cpp:25]
libUnrealEditor-Core.so!FRunnableThreadPThread::_ThreadProc(void*) [/home/rho/dev/UnrealEngine/Engine/Source/Runtime/Core/Private/HAL/PThreadRunnableThread.h:185]
libc.so.6!UnknownFunction(0x94b42)
libc.so.6!UnknownFunction(0x1269ff)`

**Comment by maumock** [comment]
**Crash**
In prior tags this feature was working, so overwrote the code block and it worked. I highly doubt this is the ideal solution, but working on my build. 

**Quick Hack**

1. Open cesium-unreal/Source/CesiumRuntime/Private/CesiumTextureUtility.cpp
2. Jump to line 774
3. Overwrite the method UTexture2D* loadTextureGameThreadPart(LoadedTextureResult* pHalfLoadedTexture) with the following:
```
UTexture2D* loadTextureGameThreadPart(LoadedTextureResult* pHalfLoadedTexture) {
  TRACE_CPUPROFILER_EVENT_SCOPE(Cesium::LoadTexture)

  if (!pHalfLoadedTexture) {
    return nullptr;
  }

  UTexture2D* pTexture = pHalfLoadedTexture->pTexture.Get();
  if (!pTexture && pHalfLoadedTexture->pTextureData) {
    pTexture = NewObject<UTexture2D>(
        GetTransientPackage(),
        NAME_None,
        RF_Transient | RF_DuplicateTransient | RF_TextExportTransient);

  #if ENGINE_MAJOR_VERSION >= 5
      pTexture->SetPlatformData(pHalfLoadedTexture->pTextureData.Release());
  #else
      pTexture->PlatformData = pHalfLoadedTexture->pTextureData.Release();
  #endif
      pTexture->AddressX = pHalfLoadedTexture->addressX;
      pTexture->AddressY = pHalfLoadedTexture->addressY;
      pTexture->Filter = pHalfLoadedTexture->filter;
      pTexture->LODGroup = pHalfLoadedTexture->group;
      pTexture->SRGB = pHalfLoadedTexture->sRGB;
      pTexture->UpdateResource();

      pHalfLoadedTexture->pTexture = pTexture;
    }

    return pTexture;
}

```

**Comment by james-yoo** [comment]
I have exactly same issue on Linux with UE 5.1

**Comment by kring** [comment]
A few different issues being reported here, I think. The `d` at the end of the lib names indicates you've built a debug version of cesium-native. That's fine if you also use a Debug build of your game project, but otherwise (you're building a Shipping or Development build of your game project) you'll get errors saying the expected .a files are missing. The solution isn't to rename the files (that may cause other problems due to differences between debug and release builds) but instead to build the Release configuration of cesium-native.

As for the InitRHI crash, we'll have to investigate that one. Do you happen to know what RHI you're using? OpenGL or Vulkan?

**Comment by kring** [comment]
Similar crash on macOS in both UE 5.0 and 5.1.

![image](https://user-images.githubusercontent.com/924374/205524573-e692bc59-ebcf-4878-b81e-f4efce5baea9.png)

@nithinp7 can you please take a look? If we can't quickly fix the new texture uploading process, we should at least disable it on these platforms and release a v1.20.1.

**Comment by kring** [comment]
v1.19 does the same on macOS, fwiw.

**Comment by maumock** [comment]
Using Vulkan. FYI: The terrain texture does not show when rolling back the method. Also crashes on final game build. Later this week I can look deeper.

**Comment by maumock** [comment]
> A few different issues being reported here, I think. The `d` at the end of the lib names indicates you've built a debug version of cesium-native. That's fine if you also use a Debug build of your game project, but otherwise (you're building a Shipping or Development build of your game project) you'll get errors saying the expected .a files are missing. The solution isn't to rename the files (that may cause other problems due to differences between debug and release builds) but instead to build the Release configuration of cesium-native.
> 
> As for the InitRHI crash, we'll have to investigate that one. Do you happen to know what RHI you're using? OpenGL or Vulkan?

Using Vulkan. FYI: The terrain texture does not show when rolling back the method. Also crashes on final game build. Later this week I can look deeper.

**Comment by JDECOMBE** [comment]
Can confirm @kring crash. Have the same problem on M1 pro (MacOS 12.3)

**Comment by kring** [comment]
@JDECOMBE we just released v1.20.1 which should fix the crash on macOS. It's not yet available on the Epic Marketplace pending Epic's usual review, but you can download from here in the meantime:
https://github.com/CesiumGS/cesium-unreal/releases
